src/savegame.cpp

```diff
@@ -279,409 +279,502 @@ void game::load_weather(std::ifstream & fin) {
+        json.member("name", i.name);
+        json.member("x", i.x);
+        json.member("y", i.y);
+        json.member("size", i.s);
         json.end_object();
-    } catch (std::string e) {
-        //debugmsg("error saving overmap: %s", e.c_str());
     }
+    json.end_array();
     fout << std::endl;
 
-    for( auto &mgv : zg ) {
-        auto &mg = mgv.second;
-        fout << "Z " << mg.type.str() << " " << mg.posx << " " << mg.posy << " " <<
-            mg.posz << " " << int(mg.radius) << " " << mg.population << " " <<
-            mg.diffuse << " " << mg.dying << " " <<
-            mg.horde << " " << mg.tx << " " << mg.ty << " " << mg.interest << std::endl;
+    json.member("roads_out");
+    json.start_array();
+    for( auto &i : roads_out ) {
+        json.start_object();
+        json.member("x", i.x);
+        json.member("y", i.y);
+        json.end_object();
     }
-    for (auto &i : cities)
-        fout << "t " << i.x << " " << i.y << " " << i.s << std::endl;
-    for (auto &i : roads_out)
-        fout << "R " << i.x << " " << i.y << std::endl;
-    for (auto &i : radios)
-        fout << "T " << i.x << " " << i.y << " " << i.strength <<
-            " " << i.type << " " << std::endl << i.message << std::endl;
-
-    for( const auto &mdata : monster_map ) {
-        fout << "M " << mdata.first.x << " " << mdata.first.y << " " << mdata.first.z <<
-            " " << mdata.second.serialize() << std::endl;
+    json.end_array();
+    fout << std::endl;
+
+    json.member("radios");
+    json.start_array();
+    for( auto &i : radios ) {
+        json.start_object();
+        json.member("x", i.x);
+        json.member("y", i.y);
+        json.member("strength", i.strength);
+        json.member("type", radio_type_names[i.type]);
+        json.member("message", i.message);
+        json.end_object();
     }
+    json.end_array();
+    fout << std::endl;
 
-    // store tracked vehicle locations and names
-    for( const auto &elem : vehicles ) {
-        int id = elem.first;
-        om_vehicle v = elem.second;
-        fout << "v " << id << " " << v.name << " " << v.x << " " << v.y << std::endl;
+    json.member("monster_map");
+    json.start_array();
+    for( auto &i : monster_map ) {
+        i.first.serialize(json);
+        i.second.serialize(json);
     }
+    json.end_array();
+    fout << std::endl;
 
-    //saving the npcs
-    for (auto &i : npcs)
-        fout << "n " << i->save_info() << std::endl;
+    json.member("tracked_vehicles");
+    json.start_array();
+    for( const auto &i : vehicles ) {
+        json.start_object();
+        json.member("id", i.first);
+        json.member("name", i.second.name);
+        json.member("x", i.second.x);
+        json.member("y", i.second.y);
+        json.end_object();
+    }
+    json.end_array();
+    fout << std::endl;
+
+    json.member("npcs");
+    json.start_array();
+    for (auto &i : npcs) {
+        json.write( i->save_info() );
```

Comment:
This doesn't work correctly. `save_info` returns a JSON string, it is written here as a single string to JSON, but the deserialize code expects a normal JsonObject.

Example (that's what is written):
```JSON
"npcs":["{\"str_cur\":1,\"str_max\":6,\"dex_cur\":5,\"dex_max\":8,\"int_cur\":3,\"int_max\":6,\"per_cur\":7,\"per_max\":
```

This is what is expected:
```JSON
"npcs":[
    { "str_cur":1,"str_max":6,"dex_cur":5,"dex_max":8,"int_cur":3,"int_max":6,"per_cur":7,
```